### PR TITLE
Support bytesize: on Strings (#10)

### PR DESCRIPTION
**Pull request by docelic**
Support `bytesize: Nil | Int | Range(Int, Int)` on Strings

### CODE REVIEW

**Comment by docelic** [comment]
Heya,

Here is the proposal for completing #10.

When deserializing and `max_size` is reached in a String without finding `\0`, it raises an error, and a corresponding test exists.

The remaining issue is how to hook this logic into ByteFormat serialization. I looked into the end of file `src/byte_format/serialize.cr`, but I don't see how to access the annotations from there. (The `_` variable is object itself, it's not Crystalizer's `Var` on which annotations exist).

Do you have any advice? Thanks.

**Comment by j8r** [comment]
Before thinking of that, let me recap of what you need now:
- set an exact size for a String
- set a max size a String, with a delimiter, can have (possibly one would want a min size too)

Note that I prefer to use the `size` word instead of `length`. I don't have any string opinion outside Crystal, but in Crystal itself `size` is used, so let's use it too here.

A the annotation API, we can have then:
- `@[Crystalizer::Field(size: 10)]`: field size is exactly 10, no delimiter.
- `@[Crystalizer::Field(size: 10..15)`: field size must be at least 10 and maximum 15, delimiter.

There could be overloads I guess, one for  `Int32` and one for `Range(Int32, Int32)`.

Finally, `Bytes` should support this too.

Worth considering: having a default upper limit, i.e default range, to prevent security issues/crashes could be a good idea - WDYT?

**Comment by j8r** [comment]
For your question, you can add the third yielded argument `_, obj, var`. Then, you'll have to add logic to `de_unionize`. If the type is a `String`, then add the argument `size: var.annotations[:size]?`.

**Comment by docelic** [comment]
> Before thinking of that, let me recap of what you need now:
>     * set an exact size for a String

Ah yes, good idea, I wasn't thinking of that.

> Note that I prefer to use the `size` word instead of `length`. I don't have any string opinion outside Crystal, but in Crystal itself `size` is used, so let's use it too here.

Right, agreed, the PR already uses `size`. I only used `length` during the discussion on the ticket.

> A the annotation API, we can have then:
>     * `@[Crystalizer::Field(size: 10)]`: field size is exactly 10, no delimiter.
>     * `@[Crystalizer::Field(size: 10..15)`: field size must be at least 10 and maximum 15, delimiter.

OK, sounds good. I guess that would work for this initial implementation, and if we ever encounter a case where a person needs the third case (exact size + delimiter), we can then expose the delimiter setting in the annotation, so users can set it as they wish.

> There could be overloads I guess, one for `Int32` and one for `Range(Int32, Int32)`.
> 
> Finally, `Bytes` should support this too.
> 
> Worth considering: having a default upper limit, i.e default range, to prevent security issues/crashes could be a good idea - WDYT?

Don't have a particular opinion. In my case the strings will always be below 1K length, but I am not sure if this is a good default or a default should be large like 1M or so.

Thanks for comments/advice, will continue work on it tomorrow.

**Comment by docelic** [comment]
I've implemented `size : Int` and `size : Range` for String deserialization.

Things remaining to do:
- Add `var` to `de_unionize`
- Support serialization as well (changes are necessary to include or avoid \0, depending on `size: Int | Range | Nil`)
- Make it work for `Bytes` as well

File: src/byte_format/serialize.cr [hunk 1]
```diff
@@ -39,22 +39,51 @@ struct Crystalizer::ByteFormat
   end
 
   # Serializes a `String` to bytes, written to the `io`, and add a trailing `string_delimiter`.
-  def serialize(string : Path | String | Symbol)
+  def serialize(string : Path | String | Symbol, add_delimiter = true)
```
Comment (j8r) [inline src/byte_format/serialize.cr]:
```suggestion
  def serialize(string : Path | String | Symbol, add_delimiter : Bool = true)
```

File: src/byte_format/deserialize.cr [hunk 2]
```diff
@@ -67,7 +67,11 @@ struct Crystalizer::ByteFormat
   end
 
   def deserialize(to type : Number::Primitive.class)
-    @io.read_bytes type, @format
+    begin
+      @io.read_bytes type, @format
+    rescue e
+      raise Error.new e.message
+    end
```
Comment (j8r) [inline src/byte_format/deserialize.cr]: Why this change?

File: src/byte_format/deserialize.cr [hunk 3]
```diff
@@ -83,6 +87,33 @@ struct Crystalizer::ByteFormat
     end
   end
 
+  # :ditto:
+  def deserialize(to type : String.class, size : Range)
```
Comment (j8r) [inline src/byte_format/deserialize.cr]:
```suggestion
  def deserialize(to type : String.class, size : Range(Int, Int))
```

File: src/byte_format/deserialize.cr [hunk 4]
```diff
@@ -83,6 +87,33 @@ struct Crystalizer::ByteFormat
...
+    end
+
+    if (min_size = size.begin) && string.size < min_size
+      raise Error.new "String size not in range: #{size}"
+    end
+
+    string
+  end
+
+  # Deserializes a `String` from reading from the `io`. String is exactly `size` bytes with no trailing '\0'.
```
Comment (j8r) [inline src/byte_format/deserialize.cr]:
```suggestion
  # Deserializes a `String` from reading from the `io`. String is exactly `size` bytes with no trailing `@string_delimiter`.
```

File: src/byte_format/deserialize.cr [hunk 5]
```diff
@@ -83,6 +87,33 @@ struct Crystalizer::ByteFormat
     end
   end
 
+  # :ditto:
+  def deserialize(to type : String.class, size : Range)
+    string = if max_size = size.end
+               # An alternative to reading limit `max_size + 1` would be to read `max_size` or `max_size - 1`
+               # (depending on `size.excludes_end?`), then peek the next char for '\0' and either consume it
+               # (if it is '\0') or set a flag for string being out of bounds.
+               @io.gets(@string_delimiter.not_nil!, max_size + 1, true) || ""
```
Comment (j8r) [inline src/byte_format/deserialize.cr]:
```suggestion
               @io.gets(@string_delimiter.as(Char), max_size + 1, true) || ""
```
Will return a better error message.

File: src/byte_format/deserialize.cr [hunk 6]
```diff
@@ -83,6 +87,33 @@ struct Crystalizer::ByteFormat
     end
   end
 
+  # :ditto:
+  def deserialize(to type : String.class, size : Range)
+    string = if max_size = size.end
+               # An alternative to reading limit `max_size + 1` would be to read `max_size` or `max_size - 1`
+               # (depending on `size.excludes_end?`), then peek the next char for '\0' and either consume it
+               # (if it is '\0') or set a flag for string being out of bounds.
```
Comment (j8r) [inline src/byte_format/deserialize.cr]: `@string_delimiter` should be mentioned instead of `\0`

File: src/byte_format/deserialize.cr (hunk 2 above)
Comment (docelic) [reply src/byte_format/deserialize.cr]: It seemed to me that you wanted to always return a local Error.new from (de)serializer modules, and without this the user would see the underlying IO::Error instead of some local Error. Do you want to keep or undo this?

**Comment by docelic** [comment]
OK, great, so at this point the only remaining issue is re. that `Error.new`. If you like my reasoning for it, you could leave it as-is and accept the commits. Or let me know and I can undo it before merging.

I would work on adding the same support for Bytes in a separate ticket. (Probably after #13.)

File: src/byte_format/deserialize.cr (hunk 2 above)
Comment (j8r) [reply src/byte_format/deserialize.cr]: This means doing this for all other methods too.
We may consider this in another PR.

Comment (j8r) [reply src/byte_format/deserialize.cr]: Also here the `cause` was missing.

Comment (docelic) [reply src/byte_format/deserialize.cr]: Ah yes, actually this was an unintended change, I intended to do this in another place, not for Number. Undo-ed it now, thanks.

File: src/byte_format/deserialize.cr [hunk 7]
```diff
@@ -83,6 +83,33 @@ struct Crystalizer::ByteFormat
...
+    if (min_size = size.begin) && string.size < min_size
+      raise Error.new "String size not in range: #{size}"
+    end
+
+    string
+  end
+
+  # Deserializes a `String` from reading from the `io`. String is exactly `size` bytes with no trailing `@string_delimiter`.
+  def deserialize(to type : String.class, size : Int)
+    @io.read_string size
```
Comment (j8r) [inline src/byte_format/deserialize.cr]: #13 is already implemented here, no?

**Comment by j8r** [comment]
As you wish for Bytes, it will be quite easy, being similar to String.

File: spec/byte_format_spec.cr [hunk 8]
```diff
@@ -91,6 +186,73 @@ describe Crystalizer::ByteFormat do
...
+    assert_byte_format_serialization(ByteFormatTest::StringGoodSize.new, bytes)
+    expect_raises Crystalizer::ByteFormat::Error, message: "String size not expected, expected: 4, have: 5" do
+      Crystalizer::ByteFormat.serialize(ByteFormatTest::StringBadSize1.new)
+    end
+    expect_raises Crystalizer::ByteFormat::Error, message: "String size not expected, expected: 6, have: 5" do
+      Crystalizer::ByteFormat.serialize(ByteFormatTest::StringBadSize2.new)
+    end
+
+    bytes = Bytes[83, 104, 111, 114, 116, 0]
+    # Serialization
```
Comment (j8r) [inline spec/byte_format_spec.cr:218]: Can be put in a `describe` block

File: spec/byte_format_spec.cr [hunk 9]
```diff
@@ -91,6 +186,73 @@ describe Crystalizer::ByteFormat do
...
+    expect_raises Crystalizer::ByteFormat::Error, message: "String size not in range: 6... (have: 5)" do
+      Crystalizer::ByteFormat.serialize(ByteFormatTest::StringBadRange5.new)
+    end
+    expect_raises Crystalizer::ByteFormat::Error, message: "String size not in range: ...4 (have: 5)" do
+      Crystalizer::ByteFormat.serialize(ByteFormatTest::StringBadRange6.new)
+    end
+    expect_raises Crystalizer::ByteFormat::Error, message: "String size not in range: 6...8 (have: 5)" do
+      Crystalizer::ByteFormat.serialize(ByteFormatTest::StringBadRange7.new)
+    end
+    # Deserialization
```
Comment (j8r) [inline spec/byte_format_spec.cr:237]: Can be put in a describe block here too

File: spec/byte_format_spec.cr (hunk 8 above)
Comment (j8r) [inline spec/byte_format_spec.cr]: Can be put in a `describe` block

File: spec/byte_format_spec.cr (hunk 9 above)
Comment (j8r) [inline spec/byte_format_spec.cr]: ditto, with `bytes = Bytes[83, 104, 111, 114, 116, 0]` inside it.

File: spec/byte_format_spec.cr [hunk 10]
```diff
@@ -91,6 +186,73 @@ describe Crystalizer::ByteFormat do
...
+    assert_byte_format_serialization(ByteFormatTest::StringGoodRange8.new, bytes)
+    assert_byte_format_serialization(ByteFormatTest::StringGoodRange9.new, bytes)
+
+    expect_raises Crystalizer::ByteFormat::Error, message: "String size not in range: 6..8" do
+      Crystalizer::ByteFormat.deserialize(bytes, to: ByteFormatTest::StringBadRange4)
+    end
+
+    # Bytes corresponding to `@a = "Short"`, no \0 at the end
+    bytes = Bytes[83, 104, 111, 114, 116]
+    assert_byte_format_serialization(ByteFormatTest::StringGoodSize.new, bytes)
```
Comment (j8r) [inline spec/byte_format_spec.cr]: Too move between Serialization and Deserialization, because used inside the latter.

File: src/byte_format/deserialize.cr (hunk 7 above)
Comment (docelic) [reply src/byte_format/deserialize.cr]: No no, this support that we added is always fixed -- it is determined by programmer at compile time. Like:

```cr
@[Crystalizer::Field(size: 10)]
```

And in #13 we're going to support it in runtime, like, when each deserialized string is of different length, and the exact length is specified in a separate field.

In other words, #13 is for cases when you're deserializing varying data like:

```
MyStruct(size_of_string: 10, content: "...string of size 10 without \0...")
MyStruct(size_of_string: 97, content: "...string of size 97 without \0...")
MyStruct(size_of_string: 47, content: "...string of size 47 without \0...")
```

See the imagined API for it that I mentioned in #13, it should make it clearer.

**Comment by docelic** [comment]
> As you wish for Bytes, it will be quite easy, being similar to String.

Yeah I already tried it, but had some trouble so need another attempt later.

File: spec/byte_format_spec.cr (hunk 8 above)
Comment (docelic) [reply spec/byte_format_spec.cr]: Good point, will do, thanks.

**Comment by docelic** [comment]
I think we should be good to go at this point!

File: src/byte_format/deserialize.cr (hunk 7 above)
Comment (j8r) [reply src/byte_format/deserialize.cr]: I don't know, I have doubts to support this natively.
I think it would be better to make serialization/deserialization configurable for this edge cases, like in the stdlib with from_json/to_json methods that can be implemented.

Then inside the type, `deserizalize` can be used directly.

File: src/byte_format/deserialize.cr [hunk 11]
```diff
@@ -83,6 +83,33 @@ struct Crystalizer::ByteFormat
     end
   end
 
+  # :ditto:
+  def deserialize(to type : String.class, size : Range(Int32?, Int32?))
```
Comment (j8r) [inline src/byte_format/deserialize.cr]:
```suggestion
  def deserialize(to type : String.class, size : Range)
```

File: src/byte_format/serialize.cr [hunk 12]
```diff
@@ -39,22 +39,51 @@ struct Crystalizer::ByteFormat
...
 
-  private def de_unionize(object : U) forall U
+  def serialize(string : String, size : Int)
+    if string.bytesize != size
+      raise Error.new "String size not expected, expected: #{size}, have: #{string.bytesize}"
+    end
+    serialize string, add_delimiter: false
+  end
+
+  def serialize(string : String, size : Range)
```
Comment (j8r) [inline src/byte_format/serialize.cr]:
```suggestion
  def serialize(string : String, size : Range(Int?, Int?))
```

File: src/byte_format/deserialize.cr (hunk 7 above)
Comment (j8r) [reply src/byte_format/deserialize.cr]: What I think, which could be included in this library, is a type which provides this functionality, instead of an annotation.

What we need is an interface, i.e. module/abstract struct - that's essentially it.

File: src/byte_format/serialize.cr (hunk 12 above)
Comment (docelic) [reply src/byte_format/serialize.cr]: > Range(Int?, Int?))

It's not possible, it says that one must use a more specific type.

File: src/byte_format/deserialize.cr (hunk 11 above)
Comment (j8r) [reply src/byte_format/deserialize.cr]: To be consistent with the serialization

**Comment by docelic** [comment]
Actually I think there's one last improvement to be done here. Instead of checking Range's `begin` and `end`, we should just check if range `includes?` the given value. Apart from being shorter to write/maintain, it will also allow one to specify things like Range().new.step(2) and so on for free (as well as very easily other types which have an `includes?` method like Arrays).

**Review by j8r — APPROVED** [review]
Great, thanks a lot @docelic!

**Comment by docelic** [comment]
By the way: we are currently using `string.size` and comparing that to the allowed values.

However, given that this is for bytes (de)serialization, maybe it is more customary to be specifying size in terms of bytes than unicode codepoints. (In other words, maybe we should be using `string.bytesize` rather than `string.size`).

Do you have any thoughts on that? I could change this as well before merging if you think it's a valid observation.

**Comment by docelic** [comment]
> maybe we should be using `string.bytesize` rather than `string.size`

I made this change right away because it seemed to make more sense. Also, when `size : Int` is passed and we call `@io.read_string size`, the `read_string()` method thinks that the value is in number of bytes, so it already defines what the value should be.

**Comment by j8r** [comment]
Ok I'm fine with this change. Then, I'd prefer to have `size` changed to `bytesize` in the other places to prevent any confusion.

**Comment by docelic** [comment]
Hm, there seems to be an issue of the current code in `serialize` not restricting types sufficiently. The source now says:

```cr
  private def de_unionize(object : U, variable : Variable) forall U
    {% for u in U.union_types %}
      if object.is_a? {{u}}
        return case variable.type
        when String.class
          if bytesize = variable.annotations.try &.[:bytesize]
            serialize object, bytesize: bytesize
          else
            serialize object
          end
        else
          serialize object
        end
      end
    {% end %}
  end
```

And I created a struct that has UInt8 and String properties, and trying to use ByteFormat on it gave:

```
There was a problem expanding macro 'macro_139671421646592'
Called macro defined in lib/crystalizer/src/byte_format/serialize.cr:64:5

 64 | {% for u in U.union_types %}

Which expanded to:

 >  1 | 
 >  2 |       if object.is_a? String
 >  3 |         return case variable.type
 >  4 |         when String.class
 >  5 |           if bytesize = variable.annotations.try &.[:bytesize]
 >  6 |             serialize object, bytesize: bytesize
 >  7 |           else
 >  8 |             serialize object
 >  9 |           end
 > 10 |         else
 > 11 |           serialize object
 > 12 |         end
 > 13 |       end
 > 14 |     
 > 15 |       if object.is_a? UInt8
 > 16 |         return case variable.type
 > 17 |         when String.class
 > 18 |           if bytesize = variable.annotations.try &.[:bytesize]
 > 19 |             serialize object, bytesize: bytesize
 > 20 |           else
 > 21 |             serialize object
 > 22 |           end
 > 23 |         else
 > 24 |           serialize object
 > 25 |         end
 > 26 |       end
 > 27 |     
Error: no overload matches 'Crystalizer::ByteFormat#serialize' with types UInt8, bytesize: Range(Nil, Int32)

Overloads are:
 - Crystalizer::ByteFormat#serialize(string : String, bytesize : Int)
 - Crystalizer::ByteFormat#serialize(string : String, bytesize : Range(Int32 | ::Nil, Int32 | ::Nil))
 - Crystalizer::ByteFormat#serialize(number : Number::Primitive)
 - Crystalizer::ByteFormat#serialize(bool : Bool)
 - Crystalizer::ByteFormat#serialize(bytes : Bytes)
 - Crystalizer::ByteFormat#serialize(object : Enum)
 - Crystalizer::ByteFormat#serialize(object : NamedTuple | Hash)
 - Crystalizer::ByteFormat#serialize(array : Array | Deque | Set | Tuple)
 - Crystalizer::ByteFormat#serialize(string : Path | String | Symbol, add_delimiter : Bool = true)
 - Crystalizer::ByteFormat#serialize(object : O) forall O
```

Do you have an idea how to restrict the types more effectively?
